Lib/test/test_httpservers.py

```diff
@@ -1536,6 +1541,246 @@ def test_server_test_ipv4(self, _):
+                    'tls_key': self.tls_key,
+                }
+                call_args = self.args | call_args
+                mock_func.assert_called_once_with(**call_args)
+                mock_func.reset_mock()
+
+    @unittest.skipIf(ssl is None, "requires ssl")
+    @mock.patch('http.server.test')
+    def test_tls_cert_and_key_and_password_flags(self, mock_func):
+        for tls_cert_option in self.tls_cert_options:
+            for tls_key_option in self.tls_key_options:
+                for tls_password_option in self.tls_password_options:
+                    self.invoke_httpd(tls_cert_option,
+                                      self.tls_cert,
+                                      tls_key_option,
+                                      self.tls_key,
+                                      tls_password_option,
+                                      self.tls_password_file)
+                    call_args = {
+                        'tls_cert': self.tls_cert,
+                        'tls_key': self.tls_key,
+                        'tls_password': self.tls_password,
+                    }
+                    call_args = self.args | call_args
+                    mock_func.assert_called_once_with(**call_args)
+                    mock_func.reset_mock()
+
+    @unittest.skipIf(ssl is None, "requires ssl")
+    @mock.patch('http.server.test')
+    def test_missing_tls_cert_flag(self, mock_func):
+        for tls_key_option in self.tls_key_options:
+            for tls_cert_option in self.tls_cert_options:
+                with self.assertRaises(SystemExit):
+                    self.invoke_httpd(tls_key_option, self.tls_key)
+                mock_func.reset_mock()
+
+    @unittest.skipIf(ssl is None, "requires ssl")
+    @mock.patch('http.server.test')
+    def test_invalid_password_file(self, mock_func):
+        non_existent_file = 'non_existent_file'
+        for tls_password_option in self.tls_password_options:
+            for tls_cert_option in self.tls_cert_options:
+                with self.assertRaises(SystemExit):
+                    self.invoke_httpd(tls_cert_option,
+                                      self.tls_cert,
+                                      tls_password_option,
+                                      non_existent_file)
+
+    @mock.patch('http.server.test')
+    def test_no_arguments(self, mock_func):
+        self.invoke_httpd()
+        mock_func.assert_called_once_with(**self.args)
+        mock_func.reset_mock()
+
+    @mock.patch('http.server.test')
+    def test_help_flag(self, _):
+        options = ['-h', '--help']
+        for option in options:
+            with self.assertRaises(SystemExit):
+                _ = self.invoke_httpd(option)
+
+    @mock.patch('http.server.test')
+    def test_unknown_flag(self, _):
+        with self.assertRaises(SystemExit):
+            _ = self.invoke_httpd('--unknown-flag')
+
+class CommandLineRunTimeTestCase(unittest.TestCase):
```

Comment:
```suggestion
            with self.assertRaises(SystemExit) as cm:
                _ = self.invoke_httpd(option)
            self.assertEqual(cm.exception.code, 0)

    @mock.patch('http.server.test')
    def test_unknown_flag(self, _):
        with self.assertRaises(SystemExit):
            _ = self.invoke_httpd('--unknown-flag')
        self.assertEqual(cm.exception.code, 2)


class CommandLineRunTimeTestCase(unittest.TestCase):
```

If you don't want to check the exit code, capture the output outside `invoke_httpd`:

```py
def invoke_httpd(self, *args, stdout=None, stderr=None):
    stdout = StringIO() if stdout is None else stdout
    stderr = StringIO() if stderr is None else stderr
    with contextlib.redirect_stdout(stdout), \
        contextlib.redirect_stderr(stderr):
        server._main(args)
    return {'stdout': stdout.getvalue(), 'stderr': stderr.getvalue()}
```

Then do:

```py
stdout = StringIO()
stderr = StringIO()
with self.assertRaises(SystemExit):
    _ = self.invoke_httpd('--unknown-flag', stdout=stdout, stderr=stderr)
self.assertEqual(stdout.getvalue(), b"")
self.assertIn(stderr.getvalue(), b"error: ")
```